modules/ethereum-bsc/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+impl ChainTime for () {
+	/// Is a header timestamp ahead of the current on-chain time.
+	///
+	/// Check whether `timestamp` is ahead (i.e greater than) the current on-chain
+	/// time. If so, return `true`, `false` otherwise.
+	fn is_timestamp_ahead(&self, timestamp: u64) -> bool {
+		// This should succeed under the contraints that the system clock works
+		let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default();
+		Duration::from_secs(timestamp) > now
+	}
+}
+
+/// Callbacks for header submission rewards/penalties.
+pub trait OnHeadersSubmitted<AccountId> {
+	/// Called when valid headers have been submitted.
+	///
+	/// The submitter **must not** be rewarded for submitting valid headers, because greedy authority
+	/// could produce and submit multiple valid headers (without relaying them to other peers) and
+	/// get rewarded. Instead, the provider could track submitters and stop rewarding if too many
+	/// headers have been submitted without finalization.
+	fn on_valid_headers_submitted(submitter: AccountId, headers: &[BSCHeader]);
+	/// Called when invalid headers have been submitted.
+	fn on_invalid_headers_submitted(submitter: AccountId);
+	/// Called when earlier submitted headers have been finalized.
+	///
+	/// finalized is the finalized authority set
+	fn on_valid_authority_finalized(submitter: AccountId, finalized: &[Address]);
+}
+
+impl<AccountId> OnHeadersSubmitted<AccountId> for () {
+	fn on_valid_headers_submitted(_submitter: AccountId, _headers: &[BSCHeader]) {}
+	fn on_invalid_headers_submitted(_submitter: AccountId) {}
+	fn on_valid_authority_finalized(_submitter: AccountId, _finalized: &[Address]) {}
+}
+
+/// The module configuration trait.
+pub trait Config: frame_system::Config {
+	/// BSC configuration.
+	type BSCConfiguration: Get<BSCConfiguration>;
+	/// Header timestamp verification against current on-chain time.
+	type ChainTime: ChainTime;
+	/// Handler for headers submission result.
+	type OnHeadersSubmitted: OnHeadersSubmitted<Self::AccountId>;
+}
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// Block number isn't sensible.
+		RidiculousNumber,
+		/// The size of submitted headers is not N/2+1
+		InvalidHeadersSize,
+		/// This header is not checkpoint,
+		NotCheckpoint,
+		/// Invalid signer
+		InvalidSigner,
+		/// Submitted headers not enough
+		HeadersNotEnough,
+		/// Signed recently
+		SignedRecently,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		/// Verify unsigned relayed headers and finalize authority set
+		#[weight = 0]
+		pub fn verify_and_update_authority_set_unsigned(origin, headers: Vec<BSCHeader>) {
+			// ensure not signed
+			frame_system::ensure_none(origin)?;
+
+			// get finalized authority set from storage
+			let last_authority_set = &FinalizedAuthority::get();
+
+			// ensure valid length
+			assert!(last_authority_set.len() / 2 < headers.len(), "Invalid headers size");
```

Comment:
**security audit** here:

Should not use `assert!` here which will cause runtime panicked, use `ensure!`